amcl/src/amcl_node.cpp

```diff
@@ -1582,47 +1574,77 @@ AmclNode::laserReceived(const sensor_msgs::LaserScanConstPtr& laser_scan)
       else
       {
         pf_vector_t pure_odom_delta, pure_amcl_delta;
-        ros::Time base_time;
         geometry_msgs::PoseWithCovarianceStamped reliable_pose_msg;
 
-        pure_amcl_delta = getAmclMovement(p, base_time, reliable_pose_msg);
-        pure_odom_delta = getOdomMovement(pose, p.header.stamp, base_time);
-        double dx = sqrt(std::pow(pure_odom_delta.v[0] - pure_amcl_delta.v[0], 2) +
-                         std::pow(pure_odom_delta.v[1] - pure_amcl_delta.v[1], 2));
-        double dz = angle_diff(pure_odom_delta.v[2], pure_amcl_delta.v[2]);
-        bool acceptable_x = std::fabs(dx) < odom_amcl_x_thre_;
-        bool acceptable_z = std::fabs(dz) < odom_amcl_z_thre_;
+        geometry_msgs::PoseWithCovarianceStamped current_base_link_pose =
+          getCurrentBaseLinkPose(laser_scan->header.stamp);
 
-        // For debug
-        std_msgs::Float32 dx_msg, dz_msg, diff_count_msg;
-        dx_msg.data = dx;
-        dz_msg.data = dz;
-        diff_count_msg.data = float(diff_count_);
-        odom_amcl_diff_x_pub_.publish(dx_msg);
-        odom_amcl_diff_z_pub_.publish(dz_msg);
-        diff_count_pub_.publish(diff_count_msg);
+        // initialize
+        if (!is_init_odom_correction_)
+        {
+          previous_base_link_pose_ = current_base_link_pose;
+          previous_amcl_pose_ = p;
+          reliable_base_link_pose_ = current_base_link_pose;
+          reliable_amcl_pose_ = p;
+          is_init_odom_correction_ = true;
+        }
+
+        // get movement
+        pure_amcl_delta = getAmclMovement(p, previous_amcl_pose_);
+        pure_odom_delta = getOdomMovement(current_base_link_pose, previous_base_link_pose_);
+
+        // update sum
+        pure_amcl_delta_sum_.v[0] += pure_amcl_delta.v[0];
+        pure_amcl_delta_sum_.v[1] += pure_amcl_delta.v[1];
+        pure_amcl_delta_sum_.v[2] += pure_amcl_delta.v[2];
+        pure_amcl_delta_sum_.v[2]  = normalize(pure_amcl_delta_sum_.v[2]);
+        pure_odom_delta_sum_.v[0] += pure_odom_delta.v[0];
+        pure_odom_delta_sum_.v[1] += pure_odom_delta.v[1];
+        pure_odom_delta_sum_.v[2] += pure_odom_delta.v[2];
+        pure_odom_delta_sum_.v[2]  = normalize(pure_odom_delta_sum_.v[2]);
+
+        double dx = sqrt(std::pow(pure_odom_delta_sum_.v[0] - pure_amcl_delta_sum_.v[0], 2) +
+                         std::pow(pure_odom_delta_sum_.v[1] - pure_amcl_delta_sum_.v[1], 2));
+        double dz = angle_diff(pure_odom_delta_sum_.v[2], pure_amcl_delta_sum_.v[2]);
```

Comment:
dxという名前が直感的にはないと思います。
何かしら y も含んでいることを表現できると良いと思います。
例えば、update_min_d, update_min_aにならって、distanceとangleから取るとか?

I don't think the name dx is intuitive.
It would be nice to be able to express that it also includes y in some way.

For example, maybe we could take it from distance and angle, like update_min_d and update_min_a?

Reply:
> 例えば、update_min_d, update_min_aにならって、distanceとangleから取るとか?

dd, da はちょっとわかりにくいので、dx, dz のほうがまだわかりやすい感じは個人的にはあります。